specs/merkle-tree.md

```diff
@@ -0,0 +1,166 @@
+
+#### Output
+Outputs a [SphinxMerkleTree](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/packages/contracts/src/merkle-tree.ts#L128) object that follows the above [architecture](#merkle-tree-architecture) and [invariants](#high-level-merkle-tree-invariants).
+
+## Dependencies
+The Merkle tree generation utility makes calls to external libraries. We use `ethers` version 6.7.0 to handle ABI encoding tree leaf `data`, and we use `@openzeppelin/merkle-tree` version 1.0.5 to assemble the Merkle tree. We test that interactions with these libraries work properly in our [SphinxModuleProxy unit tests](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/packages/contracts/test/SphinxModuleProxy.t.sol) which cover that Merkle trees generated by this logic have data that is encoded properly and that they are executable on-chain. However, we do not test the internals of these libraries and instead rely on the assumption that they are robust and bug-free.
+
+## Assumptions
+
+### Executor Behavior Regarding Invalid Merkle Trees
+The Merkle tree generation utility provided here is used to generate Merkle trees for the official Sphinx Foundry plugin, but there is not hard requirement that it be used to generate all Merkle trees. It is entirely possible that some third-party could implement their own plugin and with some alternative generation logic.
+
+Therefore, it is possible to use alternative generation logic to constuct Merkle trees that fit the high level [architecture](#merkle-tree-architecture) of the Sphinx Merkle tree, but that are not executable and are invalid according to the [invariants](#high-level-merkle-tree-invariants). For example, one could create a tree that contains multiple leaves with the same `index` and `chainId` fields. In this case, it would only be possible to execute one of these leaves. As a result, it would not be possible to fully execute this Merkle tree and furthermore it would be ambiguous exactly which leaf should be executed if one were to attempt to execute the tree.
+
+It is also possible to construct Merkle trees that are invalid according to the invariants defined above, but are still executable on-chain. For example, one could construct a Merkle tree with a different method of determining the position of leaves within the tree (such as reverse or a random order), but that still have valid `index` and `chainId` fields for every leaf. It would be possible to execute a tree like this as long as the leaves are submitted in the correct order according to their `index` field. However, this tree would not be valid according to the invariants defined above.
```

Comment:
> It would be possible to execute a tree like this as long as the leaves are submitted in the correct order according to their `index` field. However, this tree would not be valid according to the invariants defined above.

It'd be nice to be explicit about which invariant this violates